### PR TITLE
Sets clearer warning for keyby in ad-hoc columns

### LINKED ISSUE #2763: Confusing warning message
I'm trying to split data according to 15-minute timestamps; my approach is:

```
library(data.table)
set.seed(210349)
DT = data.table(t = sample(10000, 100), V = rnorm(100))
# t is in seconds, so 900 seconds is 15 minutes
DT[ , ID := .GRP, keyby = .(t %/% 900L)]
```

> Warning message:
> In `[.data.table`(DT, , `:=`(ID, .GRP), keyby = .(t%/%900L)) :
>   := keyby not straightforward character column names or list() of column names, treating as a by:t%/%900

It's weird that this gives me a warning since it's purely related to the use of `keyby`:

```
# (no warning)
DT[order(t), ID := .GRP, by = .(t %/% 900L)]
```

(`i` in this case doesn't need to match that in `by` exactly)

For sanity:

```
DT[ , ID := .GRP, keyby = .(t %/% 900L)]
DT[order(t), ID2 := .GRP, by = .(t %/% 900L)]
DT[ , identical(ID, ID2)]
# [1] TRUE
```

What's the reason for this warning?

### PR DESCRIPTION
**Pull request by eliocamp**
Hi! 
I wanted to contribute with some easy fixes to start learning about the codebase. 

This fixes #2763. 

I changed the appropriate test. I tested locally and got some errors unrelated to my change (since they were related to dates, I think they might have to do with my locale?).

### CODE REVIEW

**Comment by MichaelChirico** [comment]
Thanks for this! I'm afraid I misled in the issue about what is going on. Here's the relevant code block:

```
if (keyby) {
  cnames = as.character(bysubl)[-1L]
  cnames = gsub('^`|`$', '', cnames)  # the wrapping backticks that were added above can be removed now, #3378
  if (all(cnames %chin% names(x))) {
    if (verbose) {last.started.at=proc.time();cat("setkey() after the := with keyby= ... ");flush.console()}
    setkeyv(x,cnames)  # TO DO: setkey before grouping to get memcpy benefit.
    if (verbose) {cat(timetaken(last.started.at),"\n"); flush.console()}
  }
  else warning(":= keyby not straightforward character column names or list() of column names, treating as a by:",paste(cnames,collapse=","),"\n")
}
return(suppPrint(x))
```

So normal behavior is: columns from by are `%chin%` `names(x)`, so we can simply run `setkeyv`. If not, there's no way to `setkey` to columns that don't exist.

Hence the else branch simply throws this warning and then `return()`s `suppPrint(x)`. So this part isn't true (sorry!) "Skipping to setkey() following computation in j" since no `setkey` is done.

**Comment by MichaelChirico** [comment]
Don't forget to add an entry to `NEWS` :)

**Comment by MichaelChirico** [comment]
I'm wondering now whether this actually rises to the level of being a `warning` or is something that can just be reported in `verbose`? @jangorecki thoughts?

**Comment by eliocamp** [comment]
I was wondering the same, since the end result is exactly the same, isn't it?

**Comment by eliocamp** [comment]
I deleted the second sentence, then!

I'm the end result is not a keyed data.table, so I think it's important to at least throw a warning to alert the user that the result is not as expected.

File: R/data.table.R [hunk 1]
```diff
@@ -1756,7 +1756,7 @@ replace_order = function(isub, verbose, env) {
         setkeyv(x,cnames)  # TO DO: setkey before grouping to get memcpy benefit.
         if (verbose) {cat(timetaken(last.started.at),"\n"); flush.console()}
       }
-      else warning(":= keyby not straightforward character column names or list() of column names, treating as a by:",paste(cnames,collapse=","),"\n")
+      else warning("keyby can only be used with existing columns.")
```
Comment (MichaelChirico) [inline R/data.table.R]: Ok looking great. I would also work this part back in: `paste(cnames, collapse=",")` to help out in scenarios where maybe you have a typo, etc

Comment (eliocamp) [reply R/data.table.R]: Done!

**Comment by mattdowle** [comment]
The new warning is `keyby can only be used with existing columns, treating as by=`.

But that's not true:

```R
> DT = data.table(a=1:2, b=1:10)
> DT
        a     b
    <int> <int>
 1:     1     1
 2:     2     2
 3:     1     3
 4:     2     4
 5:     1     5
 6:     2     6
 7:     1     7
 8:     2     8
 9:     1     9
10:     2    10
> DT[,sum(a),by=b%%2L]
       b    V1
   <int> <int>
1:     1     5
2:     0    10
> DT[,sum(a),keyby=b%%2L]
       b    V1
   <int> <int>
1:     0    10
2:     1     5
> key(.Last.value)
[1] "b"
> 
```

It's `keyby=` in combination with `:=` that this warning refers to. The old warning message referred to `:=` but the new one doesn't.

**Pull request merged**